File: orc_pkg/runner.py
"""Spawn pi workers: foreground/background runs with quota gating and idle watchdog."""
import json
import os
import selectors
import signal
import subprocess
import sys
import time
from pathlib import Path

from orc_pkg import quota, registry

PI_BASE = [
    "pi",
    "-p",
    "--mode",
    "json",
    "--offline",
    "--provider",
    "minimax",
    "--model",
    "MiniMax-M3",
    "--no-session",
]


def quota_gate(force: bool) -> bool:
    q = quota.get_quota()
    level = q.get("level", "unknown")
    if level == "warn":
        print(
            f"ORC WARNING: MiniMax quota low — 5h window {q['five_hour_pct']}% "
            f"/ weekly {q['weekly_pct']}% remaining. Consider pausing delegation.",
            file=sys.stderr,
        )
        return True
    if level == "block":
        print(
            f"ORC BLOCKED: MiniMax quota below block threshold "
            f"(5h {q['five_hour_pct']}%, weekly {q['weekly_pct']}%). "
            f"Use --force to override.",
            file=sys.stderr,
        )
        return force
    if level == "unknown":
        print(
            f"ORC NOTE: quota unknown ({q.get('reason', '')}) — proceeding.",
            file=sys.stderr,
        )
        return True
    return True


def finalize(rd: Path, meta: dict, code: int, usage: dict | None = None) -> None:
    if code == 0:
        status = "done"
    elif code < 0:
        status = "killed"
    else:
        status = "failed"
    meta["status"] = status
    meta["exit_code"] = code
    meta["ended_at"] = registry.now_iso()
    task = meta.get("task", "")
    log_path = rd / "output.log"
    out_size = 0
    if log_path.exists():
        try:
            out_size = log_path.stat().st_size
        except OSError:
            out_size = 0
    tokens = meta.setdefault("tokens", {})
    tokens["estimated_total"] = (len(task) + out_size) // 4
    if usage:
        tokens.update(usage)
        tokens["estimated_total"] = usage.get("total", tokens["estimated_total"])
    registry.write_meta(rd, meta)


def _exec(rd: Path, echo: bool = False, idle_timeout: float | None = None) -> int:
    meta = registry.read_meta(rd)
    if idle_timeout is None:
        idle_timeout = float(quota.load_config().get("idle_timeout_sec", 300))

    log_file = open(rd / "output.log", "ab")
    code = 0
    buf = b""
    usage: dict | None = None
    echoed_delta = False

    def handle_line(line: bytes) -> None:
        # Raw event line goes to the log; only extracted text (or non-JSON
        # passthrough) reaches the caller's stdout.
        nonlocal usage, echoed_delta
        log_file.write(line + b"\n")
        log_file.flush()
        try:
            evt = json.loads(line)
        except ValueError:
            evt = None
        if not isinstance(evt, dict):
            if echo and line:
                sys.stdout.buffer.write(line + b"\n")
                sys.stdout.buffer.flush()
            return
        text = _extract_text(evt)
        if text and echo:
            sys.stdout.write(text)
            sys.stdout.flush()
            echoed_delta = True
        if evt.get("type") in TERMINAL_EVENTS:
            usage = _extract_usage(evt) or usage

    try:
        try:
            proc = subprocess.Popen(
                PI_BASE + [meta["task"]],
                cwd=meta["cwd"],
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                start_new_session=True,
            )
        except FileNotFoundError:
            log_file.write(b"orc: pi executable not found on PATH\n")
            log_file.flush()
            if echo:
                sys.stderr.buffer.write(b"orc: pi executable not found on PATH\n")
                sys.stderr.buffer.flush()
            finalize(rd, meta, 127)
            return 127

        meta["pid"] = proc.pid
        meta["status"] = "running"
        registry.write_meta(rd, meta)

        sel = selectors.DefaultSelector()
        sel.register(proc.stdout, selectors.EVENT_READ)
        last_output = time.monotonic()

        try:
            try:
                while True:
                    events = sel.select(timeout=0.5)
                    if not events:
                        if proc.poll() is not None:
                            break
                        if idle_timeout > 0 and (time.monotonic() - last_output) > idle_timeout:
                            msg = (
                                f"\norc: idle timeout after {int(idle_timeout)}s — killing worker\n"
                            ).encode()
                            log_file.write(msg)
                            log_file.flush()
                            if echo:
                                sys.stderr.buffer.write(msg)
                                sys.stderr.buffer.flush()
                            try:
                                os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
                            except ProcessLookupError:
                                pass
                            proc.wait()
                            finalize(rd, meta, 124, usage)
                            return 124
                        continue
                    chunk = os.read(proc.stdout.fileno(), 65536)
                    if not chunk:
                        break
                    buf += chunk
                    while b"\n" in buf:
                        line, buf = buf.split(b"\n", 1)
                        handle_line(line)
                    last_output = time.monotonic()
            except KeyboardInterrupt:
                try:
                    os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
                except ProcessLookupError:
                    pass
                code = proc.wait()
                if code >= 0:
                    code = -signal.SIGTERM
            else:
                code = proc.wait()
        finally:
            if buf:
                try:
                    handle_line(buf)
                except Exception:
                    pass
                buf = b""
            try:
                sel.close()
            except Exception:
                pass
            try:
                proc.stdout.close()
            except Exception:
                pass
    finally:
        log_file.close()

    if echoed_delta:
        sys.stdout.write("\n")
        sys.stdout.flush()
    # pi traps SIGTERM and exits 143 (128+15) instead of dying by signal; an
    # inbox kill marker or that code means "killed by orc", not "failed".
    if code > 0 and (code == 143 or _inbox_has_kill(rd)):
        code = -signal.SIGTERM
    finalize(rd, meta, code, usage)
    if code >= 0:
        return max(code, 0)
    return 130


RPC_BASE = [
    "pi",
    "--mode",
    "rpc",
    "--offline",
    "--provider",
    "minimax",
    "--model",
    "MiniMax-M3",
    "--no-session",
]
# Real pi rpc protocol (verified live 2026-07-10): assistant text arrives as
# {"type":"message_update","assistantMessageEvent":{"type":"text_delta","delta":...}}
# and the run terminates with {"type":"agent_end",...}. pi exits when stdin closes,
# so stdin must stay open until agent_end.
TERMINAL_EVENTS = {"agent_end"}


def _extract_text(evt: dict):
    ame = evt.get("assistantMessageEvent")
    if isinstance(ame, dict) and ame.get("type") == "text_delta":
        delta = ame.get("delta")
        if isinstance(delta, str):
            return delta
    return None


def _inbox_has_kill(rd: Path) -> bool:
    inbox = rd / "inbox"
    return inbox.is_dir() and any(inbox.glob("kill-*.json"))


def _killpg(proc) -> None:
    try:
        os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
    except ProcessLookupError:
        pass


def _session_for(args) -> str | None:
    return getattr(args, "session", None) or os.environ.get("ORC_SESSION") or None


def cmd_rpc(args) -> int:
    if not quota_gate(args.force):
        return 3
    rd = registry.new_run(args.task, brain=args.brain, cwd=args.cwd,
                          session=_session_for(args))
    meta = registry.read_meta(rd)
    idle_timeout = args.idle_timeout
    if idle_timeout is None:
        idle_timeout = float(quota.load_config().get("idle_timeout_sec", 300))
    killed = False
    code = 0
    with (rd / "output.log").open("ab") as log:
        try:
            proc = subprocess.Popen(
                RPC_BASE, cwd=meta["cwd"], stdin=subprocess.PIPE,
                stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                start_new_session=True)
        except FileNotFoundError:
            log.write(b"orc: pi executable not found on PATH\n")
            print("orc: pi executable not found on PATH", file=sys.stderr)
            finalize(rd, meta, 127)
            return 127
        meta["pid"] = proc.pid
        meta["status"] = "running"
        registry.write_meta(rd, meta)
        proc.stdin.write(json.dumps(
            {"type": "prompt", "message": meta["task"]}).encode() + b"\n")
        proc.stdin.flush()

        sel = selectors.DefaultSelector()
        sel.register(proc.stdout, selectors.EVENT_READ)
        last_output = time.monotonic()
        buf = b""
        done = False
        usage = None
        try:
            while not done:
                if _inbox_has_kill(rd):
                    _killpg(proc)
                    killed = True
                    break
                events = sel.select(timeout=0.3)
                if not events:
                    if proc.poll() is not None:
                        break
                    if idle_timeout > 0 and time.monotonic() - last_output > idle_timeout:
                        msg = f"\norc: idle timeout after {int(idle_timeout)}s — killing worker\n"
                        log.write(msg.encode())
                        log.flush()
                        print(msg, file=sys.stderr)
                        _killpg(proc)
                        proc.wait()
                        finalize(rd, registry.read_meta(rd), 124)
                        return 124
                    continue
                chunk = os.read(proc.stdout.fileno(), 65536)
                if not chunk:
                    break
                last_output = time.monotonic()
                buf += chunk
                while b"\n" in buf:
                    line, buf = buf.split(b"\n", 1)
                    log.write(line + b"\n")
                    log.flush()
                    try:
                        evt = json.loads(line)
                    except ValueError:
                        sys.stdout.buffer.write(line + b"\n")
                        sys.stdout.buffer.flush()
                        continue
                    text = _extract_text(evt)
                    if text:
                        sys.stdout.write(text)
                        sys.stdout.flush()
                    if evt.get("type") in TERMINAL_EVENTS:
                        usage = _extract_usage(evt)
                        done = True
        except KeyboardInterrupt:
            _killpg(proc)
            killed = True
        finally:
            sel.close()
            try:
                proc.stdin.close()
            except (OSError, BrokenPipeError):
                pass
            if killed:
                _killpg(proc)
            code = proc.wait()
    print()
    finalize(rd, registry.read_meta(rd), -signal.SIGTERM if killed else code, usage)
    return 130 if killed else max(code, 0)


def _extract_usage(evt: dict):
    """Pull real token usage from agent_end's assistant messages (pi records it)."""
    best = None
    for msg in evt.get("messages", []) or []:
        u = msg.get("usage") if isinstance(msg, dict) else None
        if isinstance(u, dict) and u.get("totalTokens"):
            best = {"input": u.get("input", 0), "output": u.get("output", 0),
                    "cache_read": u.get("cacheRead", 0),
                    "total": u.get("totalTokens", 0)}
            cost = u.get("cost")
            if isinstance(cost, dict) and cost.get("total") is not None:
                best["cost_usd"] = round(cost["total"], 6)
    return best


def cmd_run(args) -> int:
    if not quota_gate(args.force):
        return 3
    rd = registry.new_run(args.task, brain=args.brain, cwd=args.cwd,
                          session=_session_for(args))
    if args.name:
        meta = registry.read_meta(rd)
        meta["name"] = args.name
        registry.write_meta(rd, meta)
    if args.bg:
        cmd = [sys.executable, "-m", "orc_pkg", "_exec", str(rd)]
        if args.idle_timeout is not None:
            cmd.extend(["--idle-timeout", str(args.idle_timeout)])
        subprocess.Popen(
            cmd,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
        print(rd.name)
        return 0
    return _exec(rd, echo=True, idle_timeout=args.idle_timeout)


def cmd_exec(args) -> int:
    return _exec(Path(args.run_dir), echo=args.echo, idle_timeout=args.idle_timeout)

File: rust/crates/orc-core/src/runner.rs
use std::collections::HashSet;
use std::fs::{self, File, OpenOptions};
use std::io::{BufRead, BufReader, BufWriter, Write};
use std::path::{Path, PathBuf};
use std::process::{Child, ChildStdin, Command, ExitStatus, Stdio};
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::mpsc::{self, RecvTimeoutError};
use std::thread;
use std::time::{Duration, Instant};

use anyhow::{Context, Result};
use serde_json::Value;

use crate::inbox::{acknowledge_prompt, has_kill, pending_prompts, read_prompt};
use crate::model::RunMeta;
use crate::registry::{now_iso, read_meta, write_meta};

const JSON_ARGS: &[&str] = &[
    "-p",
    "--mode",
    "json",
    "--offline",
    "--provider",
    "minimax",
    "--model",
    "MiniMax-M3",
    "--no-session",
];

const RPC_ARGS: &[&str] = &[
    "--mode",
    "rpc",
    "--offline",
    "--provider",
    "minimax",
    "--model",
    "MiniMax-M3",
    "--no-session",
];

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Mode {
    Json,
    Rpc,
}

impl Mode {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Json => "json",
            Self::Rpc => "rpc",
        }
    }

    #[must_use]
    pub fn from_meta(meta: &RunMeta) -> Self {
        if meta.mode.as_deref() == Some("rpc") {
            Self::Rpc
        } else {
            Self::Json
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct Usage {
    pub input: u64,
    pub output: u64,
    pub cache_read: u64,
    pub total: u64,
    pub cost_usd: Option<f64>,
}

#[derive(Debug)]
struct LineEvent {
    bytes: Vec<u8>,
}

fn spawn_pi(meta: &RunMeta, mode: Mode) -> Result<Child> {
    let mut command = Command::new("pi");
    command
        .args(match mode {
            Mode::Json => JSON_ARGS,
            Mode::Rpc => RPC_ARGS,
        })
        .current_dir(&meta.cwd)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    if mode == Mode::Json {
        command.arg(&meta.task).stdin(Stdio::null());
    } else {
        command.stdin(Stdio::piped());
    }
    #[cfg(unix)]
    {
        use std::os::unix::process::CommandExt;
        command.process_group(0);
    }
    command.spawn().context("spawn pi")
}

fn pump_reader<R: std::io::Read + Send + 'static>(reader: R, sender: mpsc::Sender<LineEvent>) {
    thread::spawn(move || {
        let mut reader = BufReader::new(reader);
        loop {
            let mut bytes = Vec::new();
            match reader.read_until(b'\n', &mut bytes) {
                Ok(0) | Err(_) => break,
                Ok(_) => {
                    if bytes.last() == Some(&b'\n') {
                        bytes.pop();
                        if bytes.last() == Some(&b'\r') {
                            bytes.pop();
                        }
                    }
                    if sender.send(LineEvent { bytes }).is_err() {
                        break;
                    }
                }
            }
        }
    });
}

fn compact_log_line(bytes: &[u8]) -> Vec<u8> {
    let Ok(value) = serde_json::from_slice::<Value>(bytes) else {
        return bytes.to_vec();
    };
    if value.get("type").and_then(Value::as_str) == Some("message_update")
        && let Some(event) = value.get("assistantMessageEvent")
    {
        return serde_json::to_vec(&serde_json::json!({
            "type": "message_update",
            "assistantMessageEvent": event,
        }))
        .unwrap_or_else(|_| bytes.to_vec());
    }
    bytes.to_vec()
}

#[must_use]
pub fn extract_text(value: &Value) -> Option<&str> {
    let event = value.get("assistantMessageEvent")?;
    (event.get("type")?.as_str()? == "text_delta").then(|| event.get("delta")?.as_str())?
}

#[must_use]
pub fn extract_usage(value: &Value) -> Option<Usage> {
    let messages = value.get("messages")?.as_array()?;
    let mut best = None;
    for message in messages {
        let Some(usage) = message.get("usage").and_then(Value::as_object) else {
            continue;
        };
        let Some(total) = usage.get("totalTokens").and_then(Value::as_u64) else {
            continue;
        };
        if total == 0 {
            continue;
        }
        best = Some(Usage {
            input: usage.get("input").and_then(Value::as_u64).unwrap_or(0),
            output: usage.get("output").and_then(Value::as_u64).unwrap_or(0),
            cache_read: usage.get("cacheRead").and_then(Value::as_u64).unwrap_or(0),
            total,
            cost_usd: usage
                .get("cost")
                .and_then(|cost| cost.get("total"))
                .and_then(Value::as_f64)
                .map(|cost| (cost * 1_000_000.0).round() / 1_000_000.0),
        });
    }
    best
}

fn write_rpc_prompt(stdin: &mut ChildStdin, message: &str) -> Result<()> {
    serde_json::to_writer(
        &mut *stdin,
        &serde_json::json!({
            "type": "prompt",
            "message": message,
        }),
    )?;
    stdin.write_all(b"\n")?;
    stdin.flush()?;
    Ok(())
}

#[cfg(unix)]
pub fn terminate_pid(pid: u32) {
    #[allow(unsafe_code)]
    unsafe {
        libc::kill(-(pid as i32), libc::SIGTERM);
    }
}

#[cfg(not(unix))]
pub fn terminate_pid(_pid: u32) {}

#[cfg(unix)]
fn exit_code(status: ExitStatus) -> i32 {
    use std::os::unix::process::ExitStatusExt;
    status
        .code()
        .unwrap_or_else(|| -status.signal().unwrap_or(1))
}

#[cfg(not(unix))]
fn exit_code(status: ExitStatus) -> i32 {
    status.code().unwrap_or(1)
}

fn context_exhausted(bytes: &[u8]) -> bool {
    let line = String::from_utf8_lossy(bytes).to_ascii_lowercase();
    [
        "context_length_exceeded",
        "maximum context length",
        "context window exceeded",
        "context limit exceeded",
        "token limit exceeded",
        "\"finish_reason\":\"length\"",
        "\"finishreason\":\"length\"",
        "\"stop_reason\":\"max_tokens\"",
        "\"stopreason\":\"max_tokens\"",
    ]
    .iter()
    .any(|signature| line.contains(signature))
}

fn finalize(
    run_dir: &Path,
    code: i32,
    usage: Option<&Usage>,
    failure_kind: Option<&str>,
) -> Result<()> {
    let mut meta = read_meta(run_dir)?;
    meta.status = if code == 0 {
        "done"
    } else if code < 0 {
        "killed"
    } else {
        "failed"
    }
    .to_owned();
    meta.exit_code = Some(code);
    meta.ended_at = Some(now_iso());
    let log_size = fs::metadata(run_dir.join("output.log")).map_or(0, |m| m.len());
    meta.tokens.estimated_total = ((meta.task.len() as u64) + log_size) / 4;
    if let Some(usage) = usage {
        meta.tokens.input = Some(usage.input);
        meta.tokens.output = Some(usage.output);
        meta.tokens.cache_read = Some(usage.cache_read);
        meta.tokens.total = Some(usage.total);
        meta.tokens.cost_usd = usage.cost_usd;
        meta.tokens.estimated_total = usage.total;
    }
    if code == 124 {
        meta.attention = Some("handoff_needed".to_owned());
        meta.failure_kind = Some("idle_timeout".to_owned());
    } else if let Some(failure_kind) = failure_kind {
        meta.attention = Some("handoff_needed".to_owned());
        meta.failure_kind = Some(failure_kind.to_owned());
    }
    write_meta(run_dir, &meta)?;
    crate::notification::run_finished(&meta);
    Ok(())
}

fn deliver_pending_prompts(
    run_dir: &Path,
    stdin: &mut ChildStdin,
    delivered: &mut HashSet<PathBuf>,
) -> Result<()> {
    for path in pending_prompts(run_dir, delivered)? {
        let message = read_prompt(&path)?;
        if message.kind != "prompt" {
            delivered.insert(path);
            continue;
        }
        write_rpc_prompt(stdin, &message.message)?;
        acknowledge_prompt(run_dir, &path)?;
        delivered.insert(path);
    }
    Ok(())
}

pub fn execute(run_dir: &Path, echo: bool, idle_timeout: f64) -> Result<i32> {
    let mut meta = read_meta(run_dir)?;
    let mode = Mode::from_meta(&meta);
    let log = OpenOptions::new()
        .create(true)
        .append(true)
        .open(run_dir.join("output.log"))?;
    let mut log = BufWriter::new(log);
    let mut child = match spawn_pi(&meta, mode) {
        Ok(child) => child,
        Err(error) => {
            let message = "orc: pi executable not found on PATH\n";
            log.write_all(message.as_bytes())?;
            log.flush()?;
            if echo {
                eprint!("{message}");
            }
            finalize(run_dir, 127, None, None)?;
            return if error.downcast_ref::<std::io::Error>().is_some() {
                Ok(127)
            } else {
                Err(error)
            };
        }
    };
    let child_pid = child.id();
    meta.pid = Some(child_pid);
    meta.status = "running".to_owned();
    write_meta(run_dir, &meta)?;

    let mut stdin = child.stdin.take();
    if mode == Mode::Rpc
        && let Some(stdin) = stdin.as_mut()
    {
        write_rpc_prompt(stdin, &meta.task)?;
    }

    let (sender, receiver) = mpsc::channel();
    if let Some(stdout) = child.stdout.take() {
        pump_reader(stdout, sender.clone());
    }
    if let Some(stderr) = child.stderr.take() {
        pump_reader(stderr, sender);
    }

    let interrupted = Arc::new(AtomicBool::new(false));
    signal_hook::flag::register(signal_hook::consts::SIGINT, Arc::clone(&interrupted))?;
    let mut delivered = HashSet::new();
    let mut last_output = Instant::now();
    let mut usage = None;
    let mut killed = false;
    let mut timed_out = false;
    let mut saw_delta = false;
    let mut exhausted_context = false;
    let poll = Duration::from_millis(if mode == Mode::Rpc { 300 } else { 500 });

    loop {
        if interrupted.load(Ordering::Relaxed) {
            terminate_pid(child_pid);
            killed = true;
            break;
        }
        if mode == Mode::Rpc {
            if has_kill(run_dir) {
                terminate_pid(child_pid);
                killed = true;
                break;
            }
            if let Some(stdin) = stdin.as_mut() {
                deliver_pending_prompts(run_dir, stdin, &mut delivered)?;
            }
        }
        match receiver.recv_timeout(poll) {
            Ok(line) => {
                last_output = Instant::now();
                exhausted_context |= context_exhausted(&line.bytes);
                let compact = compact_log_line(&line.bytes);
                log.write_all(&compact)?;
                log.write_all(b"\n")?;
                log.flush()?;
                match serde_json::from_slice::<Value>(&line.bytes) {
                    Ok(value) => {
                        if let Some(text) = extract_text(&value) {
                            if echo {
                                print!("{text}");
                                std::io::stdout().flush()?;
                            }
                            saw_delta = true;
                        }
                        if value.get("type").and_then(Value::as_str) == Some("agent_end") {
                            usage = extract_usage(&value).or(usage);
                            if mode == Mode::Rpc {
                                break;
                            }
                        }
                    }
                    Err(_) if echo && !line.bytes.is_empty() => {
                        println!("{}", String::from_utf8_lossy(&line.bytes));
                    }
                    Err(_) => {}
                }
            }
            Err(RecvTimeoutError::Timeout) => {
                if let Some(status) = child.try_wait()? {
                    let mut code = exit_code(status);
                    if code == 143 || has_kill(run_dir) {
                        code = -libc::SIGTERM;
                    }
                    if saw_delta && echo {
                        println!();
                    }
                    finalize(
                        run_dir,
                        code,
                        usage.as_ref(),
                        exhausted_context.then_some("context_exhausted"),
                    )?;
                    return Ok(if code < 0 { 130 } else { code });
                }
                if idle_timeout > 0.0
                    && last_output.elapsed() > Duration::from_secs_f64(idle_timeout)
                {
                    let message = format!(
                        "\norc: idle timeout after {}s — killing worker\n",
                        idle_timeout as u64
                    );
                    log.write_all(message.as_bytes())?;
                    log.flush()?;
                    if echo {
                        eprint!("{message}");
                    }
                    terminate_pid(child_pid);
                    timed_out = true;
                    break;
                }
            }
            Err(RecvTimeoutError::Disconnected) => break,
        }
    }

    drop(stdin);
    if killed || timed_out {
        terminate_pid(child_pid);
    }
    let status = child.wait()?;
    let code = if timed_out {
        124
    } else if killed || exit_code(status) == 143 || has_kill(run_dir) {
        -libc::SIGTERM
    } else {
        exit_code(status)
    };
    if saw_delta && echo {
        println!();
    }
    finalize(
        run_dir,
        code,
        usage.as_ref(),
        exhausted_context.then_some("context_exhausted"),
    )?;
    Ok(if code < 0 { 130 } else { code })
}

#[must_use]
pub fn default_output_log(run_dir: &Path) -> PathBuf {
    run_dir.join("output.log")
}

pub fn create_empty_log(run_dir: &Path) -> Result<File> {
    File::create(default_output_log(run_dir)).context("create output.log")
}

pub fn spawn_background(run_dir: &Path, idle_timeout: Option<f64>) -> Result<()> {
    let executable = std::env::current_exe().context("locate Rust orc binary")?;
    let mut command = Command::new(executable);
    command
        .arg("_exec")
        .arg(run_dir)
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null());
    if let Some(timeout) = idle_timeout {
        command.args(["--idle-timeout", &timeout.to_string()]);
    }
    #[cfg(unix)]
    {
        use std::os::unix::process::CommandExt;
        command.process_group(0);
    }
    command.spawn().context("launch background runner")?;
    Ok(())
}
